core/primitives/src/time.rs

```diff
@@ -0,0 +1,122 @@
+use std::default::Default;
+
+use chrono;
+
+pub use chrono::Utc;
+pub use std::time::{Duration, Instant};
+
+use chrono::DateTime;
+use std::cell::RefCell;
+use std::collections::VecDeque;
+
+pub struct MockClockPerThread {
+    utc: VecDeque<DateTime<Utc>>,
+    durations: VecDeque<Duration>,
+    utc_call_count: u64,
+    instant_call_count: u64,
+    instant: Instant,
+    is_mock: bool,
+}
+
+pub struct Clock {}
+
+impl MockClockPerThread {
+    pub fn reset(&mut self) {
+        self.utc.clear();
+        self.durations.clear();
+        self.utc_call_count = 0;
+        self.instant_call_count = 0;
+        self.instant = Instant::now();
+        self.is_mock = false;
+    }
+
+    fn with<F, T>(f: F) -> T
+    where
+        F: FnOnce(&mut MockClockPerThread) -> T,
+    {
+        thread_local! {
+            static INSTANCE: RefCell<MockClockPerThread> = RefCell::default()
+        }
+        INSTANCE.with(|it| f(&mut *it.borrow_mut()))
+    }
+
+    fn pop_utc(&mut self) -> Option<DateTime<chrono::Utc>> {
+        self.utc_call_count += 1;
+        self.utc.pop_front()
+    }
+    fn pop_instant(&mut self) -> Option<Instant> {
+        self.instant_call_count += 1;
+        let x = self.durations.pop_front();
+        match x {
+            Some(t) => self.instant.checked_add(t),
+            None => None,
+        }
+    }
+}
+
+impl Default for MockClockPerThread {
+    fn default() -> Self {
+        Self {
+            utc: VecDeque::with_capacity(16),
+            durations: VecDeque::with_capacity(16),
+            utc_call_count: 0,
+            instant_call_count: 0,
+            instant: Instant::now(),
+            is_mock: false,
+        }
+    }
+}
+
+impl Clock {
+    pub fn add_utc(&mut self, mock_date: DateTime<chrono::Utc>) {
+        MockClockPerThread::with(|clock| {
+            clock.utc.push_back(mock_date);
+            clock.is_mock = true;
+        });
+    }
+
+    pub fn add_instant(&mut self, mock_instant: Duration) {
+        MockClockPerThread::with(|clock| {
+            clock.durations.push_back(mock_instant);
+            clock.is_mock = true;
+        });
+    }
+
+    pub fn utc(&mut self) -> DateTime<chrono::Utc> {
```

Comment:
How about we change signature to `pub fn utc_now() -> DateTime<chrono::Utc>`

- self is not used - This will allow us to use `Clock::utc_now()` without having to create `Clock {}`
- changing `utc` to `utc_now` should make code more readable. Without prior knowledge we will know that we are getting current time.